ckeygen.1
.TH ckeygen 1 "1.0" "Internet Connection, Inc." "Ad-Hoc Clusters"
.SH NAME
ckeygen \- generate keypairs for cluster applications
.SH SYNOPSIS
.B ckeygen [-stdin] appid [appid.pub]

Generates a new private and public keypair called
.B appid
and
.B appid.pub
respectively.
.SH DESCRIPTION
Every clustered application is identified on a network by its keypair.
Both parts of the keypair are considered secret- any host/user allowed
access to the public-key portion can create work to do, and any host/user
with the private key can provide answers.

.I cservice(1)
uses the private key and serves an application associated with it
to the cluster.
.I cexec(1)
uses the public key to locate a running
.I cservice(1)
with that private key and to authenticate it.

If
.B appid
already exists, a public key will be created for it (if possible).
.SH OPTIONS
.B -stdin
will cause ckeygen to generate a private key based on the SHA-1 hash
of standard input. This can be useful in generating keys to a version.
.SH EXIT CODES
.B ckeygen
will exit 101 if the private key exists, and is invalid.

.B ckeygen
will exit 102 if it cannot create/write to the public key.

.B ckeygen
will exit 103 if it cannot create/write to the private key.

.B ckeygen
will exit 104 if generating a key from stdin and it doesn't match the
already existing private key that is supplied on the commandline.

.B ckeygen
will exit 111 in the face of temporary errors (like allocating memory).

.B ckeygen
will exit 255 if there was a problem emitting its help text.

.B ckeygen
will exit 0 if everything (otherwise) completed successfully.
.SH SEE ALSO
cexec(1), cservice(1)


CHANGELOG
2008-08-07 18:02  geocar

	* Makefile, bio.c, build-cross.sh, cexec.c, ckeygen.c, cproxy.c,
	  crat.c, cservice.c, deploy.sh, hash.c, keyfile.c, load.c,
	  multipipe.c, netnum.c, netnum.h, parity.c, sha1.c, sha1.h,
	  strline.c, strline.h, types.h: port to 64bit system

2007-03-02 22:05  geocar

	* deploy.sh: amd64

2007-01-30 21:27  geocar

	* cexec.c: wee

2006-12-08 22:48  geocar

	* bio.c, cexec.c, cservice.c, deploy.sh: moving it around...

2006-12-08 17:48  geocar

	* bio.c, build-cross.sh, cexec.c, cio.c, cservice.c, deploy.sh: fix
	  base when doing copyout; dont close on write (sigpipe helps
	  noone) and sparc is dead. oh, and nul-group is all-group

2006-07-31 06:46  geocar

	* cservice.c: support $BOOST mode

2006-07-31 06:46  geocar

	* load.c: use gnu load

2006-06-23 06:59  geocar

	* cexec.c: dont regenerate parity (we cannot remember long enough)

2006-06-23 06:59  geocar

	* cservice.c: fix delays, and make pubkey toplevel so we can test

2006-06-23 05:27  geocar

	* Makefile: uses sockop

2006-06-23 05:27  geocar

	* cexec.c: add testing feature, and skipto to try next-best host

2006-06-23 05:26  geocar

	* cservice.c: check signature for safety and use sockop_nonblock

2006-06-23 05:26  geocar

	* hash.c: zero s1/s2 barrels

2006-06-23 05:26  geocar

	* keyfile.c: check zero length vlpoint

2006-06-23 05:26  geocar

	* cio.c, sockop.c: bleh, convenience i suppose

2005-11-08 07:51  geocar

	* cexec.c: remove did_io optimization; it's complicated

2005-03-17 10:07  geocar

	* ckeygen.1: adding documentation about -stdin flag

2005-03-17 10:06  geocar

	* ckeygen.c: generate keys based on stdin with -stdin option now...

2005-03-17 10:06  geocar

	* cexec.c: did_io when writing too...

2005-03-01 12:58  geocar

	* cservice.c: memalloc isn't fatal PEER_vars, don't attempt to dup2
	  a -1 fd (bitmap)

2005-03-01 11:31  geocar

	* cproxy.c: don't waste file descriptors all wily-nilly, also
	  multihomed SO_BINDTODEVICE capable boxes are now assisted...

2005-03-01 11:31  geocar

	* cexec.c: put lseek hack back in

2005-02-17 08:11  geocar

	* parity.c: use urandom instead of random; the lifetime of the
	  check-bits are so short that other attacks are more likely to be
	  useful

2005-02-17 07:59  geocar

	* cservice.c: do bit flipping if near or far is running on a really
	  strange platform

2005-02-17 07:58  geocar

	* cexec.c: return didio hack

2005-02-17 07:58  geocar

	* cexec.c, cluster.h: move bytesex constant out of cexec

2005-02-17 07:57  geocar

	* bio.c: sink output in blast and copy

2005-02-17 06:27  geocar

	* LICENSE, Makefile, cexec.c, cio.c, cio.h, cservice.c, keyfile.c:
	  moved cread/cwrite out of keyfile and changed write-hack to use
	  4bytes instead of a vlpoint. the write-hack might prove useful
	  someday...

2005-02-16 14:40  geocar

	* cexec.c, cservice.c: turn keepalive back on

2005-02-16 13:02  geocar

	* cexec.c, cservice.c: don't use keepalive anymore, protect
	  main-channel for sanity, and add an extra vlpoint-exchange just
	  to force data both ways on the socket before working/forking

2005-02-16 13:00  geocar

	* bio.c: protect fcntl changes

2005-02-16 12:50  geocar

	* bio.c: turn off nonblock when blasting, and honor EAGAIN on
	  bio_more

2005-02-16 12:47  geocar

	* keyfile.c: turn off nonblock when in cread/cwrite

2005-02-16 09:54  geocar

	* cservice.1, cservice.c: set $REMOTE_ADDR, $REMOTE_PORT,
	  $SERVER_ADDR, $SERVER_PORT so cservice child can do things like
	  ident-check the cexec peer if it likes

2005-02-16 09:53  geocar

	* crat.c: unknown keys are listed in hex for brevity

2005-02-16 09:53  geocar

	* strline.c, strline.h: support for hex numbers too

2005-02-16 08:29  geocar

	* LICENSE, Makefile, cexec.c, crat.c, cservice.c, fdm.c, pair.c,
	  sockop.c, sockop.h: move sockops into a separate file

2005-02-16 08:28  geocar

	* cproxy.c: bugfix; detect hangup on write; move expires operation
	  before gathering fds so cproxy doesn't bomb

2005-02-15 10:44  geocar

	* cexec.c: restart processing if we haven't done any I/O but still
	  got an error,

2005-02-15 10:44  geocar

	* bio.c: treat EINTR and EAGAIN the same

2005-02-15 10:43  geocar

	* cservice.c: map fd masks correctly; increased timeouts, don't try
	  and read from wronly fds, use PIPE_BUF for buffers, and send
	  close alerts for write-errors

2005-02-15 10:42  geocar

	* fdm.c: map rdonly file descriptors correctly now; treat ttys
	  specially

2005-02-15 10:41  geocar

	* keyfile.c: treat EINTR and EAGAIN the same; also be pendantic
	  about short != 16 bits

2005-02-14 17:09  geocar

	* cexec.c: use a different (better?) announcement schedule

2005-02-14 17:09  geocar

	* cproxy.c: un-nonblock fds, turn on keepalive/ndelay on both
	  sides, honor EAGAIN

2005-02-14 17:09  geocar

	* cservice.1, cservice.c: lower default concurrency, drop
	  unnecessary checks, and recover from autorat failures faster

2005-02-14 16:35  geocar

	* load.c: system_load() isn't required to actually function

2005-02-14 11:01  geocar

	* cexec.1, ckeygen.1, cproxy.1, crat.1, cservice.1: adding
	  documentation

2005-02-14 08:57  geocar

	* cexec.c: don't waste an fd

2005-02-14 08:47  geocar

	* cexec.c, cservice.c: added pingpong detection, configurable
	  timeouts, configurable port ranges, using simplexes instead of
	  duplexes where possible, sending fd usage map, and debugging
	  nonsense for cexec

2005-02-14 08:43  geocar

	* crat.c, strline.c, strline.h: move string opers stuff

2005-02-14 08:42  geocar

	* fdm.c: fixes in detecting file descriptors on Linux

2005-02-14 08:32  geocar

	* pair.c: turn in so_linger on our pair links.

2005-02-14 08:30  geocar

	* parity.c: use /dev/random if available

2004-12-16 07:21  geocar

	* cservice.c: moving waitpid chunk earlier,change autorat timeouts,
	  and disable need_delay after delay (procrastinating autorats)

2004-12-16 06:37  geocar

	* cservice.c: default maxchildren now 100

2004-12-16 06:31  geocar


ckeygen.c
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>

#include "sha1.h"
#include "keyfile.h"

static void usage(int e)
{
	int i, r;
#define M "Usage: ckeygen [-stdin] app [app.pub]\n"
	for (i = 0; i < sizeof(M)-1;) {
		do {
			r = write(2, M+i,(sizeof(M)-1)-i);
		} while (r == -1 && errno == EINTR);
		if (r < 1) exit(255);
		i += r;
	}
	exit(e);
#undef M
}

int main(int argc, char *argv[])
{
	vl_point sec, pub, w;
	word16 kk;
	char *sf, *pf;
	int fd, i, j;
	int use_stdin = 0;

	parity_init();
	gf_init();

	if (argc > 1 && strcmp(argv[1], "-stdin") == 0) {
		use_stdin = 1;
		argv++;
		argc--;
	}


	if (argc == 1 || argc > 3) usage(0);

	if (use_stdin) {
		struct SHA1_CONTEXT c;
		SHA1_DIGEST dg;
		unsigned int total;
		char buffer[4096];

		vl_clear(pub);
		sha1_context_init(&c);
		for (total = 0;;) {
			do {
				i = read(0, buffer, sizeof(buffer));
			} while (i == -1 && errno == EINTR);
			if (i == 0) break;
			if (i < 1) exit(255);
			sha1_context_hashstream(&c, buffer, i);
			total += i;
		}
		sha1_context_endstream(&c, total);
		sha1_context_digest(&c, dg);
		w[0] = 1;
		for (i = 0; i < sizeof(SHA1_DIGEST); i++) {
			w[1] = dg[i];
			vl_shortlshift(pub, 8);
			vl_add(pub, w);
		}
		/* k... digest turned into key */
	}

	fd = open(argv[1], O_RDONLY);
	if (fd == -1) {
		unsigned int bytes;

		vl_clear(sec);
		if (use_stdin) {
			/* copy */
			for (i = 0; i <= pub[0]; i++) {
				sec[i] = pub[i];
			}
		} else {
			sec[0] = VL_UNITS-1;

			fd = open("/dev/random", O_RDONLY);
			if (fd == -1) {
				w[0] = 1;
				for (i = 0; i < sec[0]; i++) {
					vl_shortlshift(sec, 16);
					kk = ((unsigned int)parity())
						| (((unsigned int)parity()) << 1)
						| (((unsigned int)parity()) << 2)
						| (((unsigned int)parity()) << 3)
						| (((unsigned int)parity()) << 4)
						| (((unsigned int)parity()) << 5)
						| (((unsigned int)parity()) << 6)
						| (((unsigned int)parity()) << 7)
						| (((unsigned int)parity()) << 8)
						| (((unsigned int)parity()) << 9)
						| (((unsigned int)parity()) << 10)
						| (((unsigned int)parity()) << 11)
						| (((unsigned int)parity()) << 12)
						| (((unsigned int)parity()) << 13)
						| (((unsigned int)parity()) << 14)
						| (((unsigned int)parity()) << 15);
					memcpy(&w[1], &kk, sizeof(word16));
					vl_add(sec, w);
				}
			} else {
				bytes = VL_UNITS * sizeof(word16);
				for (i = 0; i < bytes;) {
					j = read(fd, (&sec[1])+i, bytes - i);
					if (j > 0) i += j; 
				}
				(void)close(fd);
			}
		}
	} else if (!vlpoint_read(sec, fd)) {
		exit(101);
	} else {
		if (use_stdin) {
			if (pub[0] != sec[0]) exit(104);
			for (i = 1; i <= pub[0]; i++)
				if (pub[i] != sec[i]) exit(104);
			/* kay, they match */
		}
		(void)close(fd);
	}
	
	sf = argv[1];
	if (argc == 2) {
		pf = (char *)malloc(strlen(sf) + 4);
		if (!pf) exit(111);
		strcpy(pf, sf);
		strcat(pf, ".pub");
	} else {
		pf = argv[2];
	}

	vl_clear(pub);
	cp_makepublickey(pub, sec);

	fd = open(pf, O_WRONLY|O_CREAT, 0666);
	if (fd == -1) exit(102);
	if (!vlpoint_write(pub, fd)) exit(102);
	close(fd);

	fd = open(sf, O_WRONLY|O_CREAT, 0666);
	if (fd == -1) exit(103);
	if (!vlpoint_write(sec, fd)) exit(103);
	close(fd);

	exit(0);
}
